src/librustc/hir/lowering.rs

```diff
@@ -1566,16 +1558,25 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
                 // to:
                 //
                 // {
-                //     match <expr> {
+                //     match { Carrier::translate( { <expr> } ) } {
```

Comment:
I'm worried about this part of the desugaring. `Carrier::translate` consumes the value, so the `Result`-ish-thing is completely destroyed and then rebuilt when leaving the match statement. This precludes writing a carrier that would like to carry something more than the success/error values (for example, a backtrace). I am wondering if there is a way we can thread some more user-configurable state through the question marks.